frontend/subrequests/lint/lint.go

```diff
@@ -50,6 +50,9 @@ func (w *Warning) PrintTo(wr io.Writer, sources []*pb.SourceInfo) error {
 		return nil
 	}
 	sourceInfo := sources[w.Location.SourceIndex]
+	if dockerfilePath != "" {
+		sourceInfo.Filename = dockerfilePath
```

Comment:
Afaics the correct filename is already filled in here:
```
docker buildx build -f app.Dockerfile  --check .
...
WARNING: ConsistentInstructionCasing - https://docs.docker.com/go/dockerfile/rule/consistent-instruction-casing/
Command 'from' should match the case of the command majority (uppercase)
app.Dockerfile:1
--------------------
   1 | >>> from alpine
   2 |     ENV foo=bar
   3 |
--------------------
```

The problem in https://github.com/docker/buildx/issues/2542 is that even though client side path is set, that path is never sent to buildkit. For example it gets replaced by temp path or subpath in https://github.com/docker/buildx/blob/96eb69aea4e03baf516f29ec36aa93a9c284f085/build/opt.go#L445-L474 . The only way to get back the original path in buildx would be to restore the path to the original form there after buildkit has returned the rebased path in response.

Reply:
Right - this is just an update that adds a variable to overwrite the default name that's printed.

Currently, `buildx` uses the `PrintLintViolations` function, which just unmarshal's the json and calls `PrintTo`. This PR allows use to continue using these functions (which provide a consistent way of printing out this warnings) while also being able to provide the reconstructed path so that we can augment the output on that.

You're absolutely right that the correct filename is filled in in that example. But for something like below, we'll see the exact same output as your example, which is the case that we're trying to address.

```
docker buildx build -f test/subdir/app.Dockerfile  --check .
...
WARNING: ConsistentInstructionCasing - https://docs.docker.com/go/dockerfile/rule/consistent-instruction-casing/
Command 'from' should match the case of the command majority (uppercase)
app.Dockerfile:1
--------------------
   1 | >>> from alpine
   2 |     ENV foo=bar
   3 |
--------------------
```

The intent of this change is to allow the user of `PrintTo` (in this case `buildx`) to provide a path which overwrites `sourceInfo.Filename` for the purposes of printing, so that we can do something along the lines of:

`warning.PrintTo(stdout, sources, "")` when the path provided to the buildkit is sufficient, but `warning.PrintTo(stdout, sources, "reconstructed/path/DockerfileName")` when there is a need to replace the printed filename with something more specific.

Reply:
Would it be better to rename `dockerfilePath` to `dockerfilePrintPathOverride` or something else more descriptive to be clear that this is the intent?

Reply:
Got it. In that case, I think the parameter should be map with mappings from buildkit filenames to the new filenames. So instead of rewriting filename for any item in `[]pb.SourceInfo` array, it would look for `app.Dockerfile` and rewrite it with `test/subdir/app.Dockerfile`. Another way would be to pass a callback with `func(*pb.SourceInfo) *pb.SourceInfo` that would be called before printing, and then buildx can pass a function that would do the swapping of Dockerfile path when it find matching candidate.

Reply:
Excellent! Updated to go with the callback option!